File: leetcode/118.pascals-triangle.js
/*
 * @lc app=leetcode id=118 lang=javascript
 *
 * [118] Pascal's Triangle
 *
 * https://leetcode.com/problems/pascals-triangle/description/
 *
 * algorithms
 * Easy (44.21%)
 * Total Accepted:    223.5K
 * Total Submissions: 505.5K
 * Testcase Example:  '5'
 *
 * Given a non-negative integer numRows, generate the first numRows of Pascal's
 * triangle.
 * 
 * 
 * In Pascal's triangle, each number is the sum of the two numbers directly
 * above it.
 * 
 * Example:
 * 
 * 
 * Input: 5
 * Output:
 * [
 * ⁠    [1],
 * ⁠   [1,1],
 * ⁠  [1,2,1],
 * ⁠ [1,3,3,1],
 * ⁠[1,4,6,4,1]
 * ]
 * 
 * 
 */
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function(numRows) {
    const arr = new Array(numRows);

    pascalsTriangle(0, numRows, arr);

    return arr;

    function pascalsTriangle(i, n, arr) {
        if (i === n) {
            return;
        }

        arr[i] = [];
        for (let j = 0; j < i + 1; j++) {
            if (0 === j || i === j) {
                arr[i][j] = 1;
            } else {
                arr[i][j] = arr[i - 1][j] + arr[i - 1][ j - 1];
            }
        }

        pascalsTriangle(i + 1, n, arr);
    }
};

File: leetcode/118.pascals-triangle.java
/*
 * @lc app=leetcode id=118 lang=java
 *
 * [118] Pascal's Triangle
 */

// @lc code=start
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> list = new ArrayList<>();
        generate(list, numRows);
        return list;
    }

    private void generate(List<List<Integer>> list, int numRows) {
        if (numRows == 1) {
            list.add(Arrays.asList(1));
        } else if (numRows > 1) {
            generate(list, numRows - 1);

            List<Integer> preList = list.get(numRows - 2);
            List<Integer> curList = new ArrayList<>();
            for (int i = 0; i < preList.size(); i++) {
                if (i == 0) {
                    curList.add(1);
                }
                
                if (i > 0) {
                    curList.add(i, preList.get(i - 1) + preList.get(i));
                }

                if (i == preList.size() - 1) {
                    curList.add(1);
                }
            }
            list.add(curList);
        }
    }
}
// @lc code=end
